File: 1857. Largest Color Value in a Directed Graph/Solution.py
class Solution:
    def largestPathValue(self, colors: str, edges: List[List[int]]) -> int:
        n, res, processed = len(colors), 0, 0
        dp = [[0 for i in range(26)] for j in range(n)]
        graph = [[] for i in range(n)]
        indeg = [0 for i in range(n)]
        q = []
        for x, y in edges:
            graph[x].append(y)
            indeg[y] += 1
        for i in range(n):
            if indeg[i] == 0: q.append(i)
        while q:
            l = len(q)
            for i in range(l):
                curr = q.pop(0)
                processed += 1
                dp[curr][ord(colors[curr]) - ord('a')] += 1
                res = max(res, dp[curr][ord(colors[curr]) - ord('a')])
                for j in graph[curr]:
                    for k in range(26):
                        dp[j][k] = max(dp[j][k], dp[curr][k])
                    indeg[j] -= 1
                    if indeg[j] == 0: q.append(j)
        return -1 if processed < n else res

File: 1857. Largest Color Value in a Directed Graph/Solution.java
class Solution {
    public int largestPathValue(String colors, int[][] edges) {
        int n = colors.length(), res = 0, processed = 0;
        int[][] dp = new int[n][26];
        List<Integer>[] graph = new List[n];
        int[] indeg = new int[n];
        Queue<Integer> q = new LinkedList<>();
        for (int i = 0; i < n; i++) graph[i] = new ArrayList<>();
        for (int[] e : edges) {
            graph[e[0]].add(e[1]);
            indeg[e[1]]++;
        }
        for (int i = 0; i < n; i++) {
            if (indeg[i] == 0) q.offer(i);
        }
        while (!q.isEmpty()) {
            int l = q.size();
            for (int i = 0; i < l; i++) {
                int curr = q.poll();
                processed++;
                dp[curr][colors.charAt(curr) - 'a']++;
                res = Math.max(res, dp[curr][colors.charAt(curr) - 'a']);
                for (int j : graph[curr]) {
                    for (int k = 0; k < 26; k++) {
                        dp[j][k] = Math.max(dp[j][k], dp[curr][k]);
                    }
                    indeg[j]--;
                    if (indeg[j] == 0) q.offer(j);
                }
            }
        }
        return processed < n ? -1 : res;
    }
}